File: Dynamic Programming/DP on Grids/Minimum Path Sum/memoization.cpp
class Solution {
    public:
        int m, n;
        vector<vector<int>> dp;
        int solve(int row, int col, vector<vector<int>>& grid) {
            if(row > m-1 || col > n-1)
                return 1e9;
    
            if(row == m-1 && col == n-1){
                return grid[row][col];
            }
    
            if(dp[row][col] != -1)
                return dp[row][col];
    
            int right = grid[row][col] + solve(row, col+1, grid);
            int down  = grid[row][col] + solve(row+1, col, grid);
    
            return dp[row][col] = min(right, down);
        }
    
        int minPathSum(vector<vector<int>>& grid) {
            m = grid.size();
            n = grid[0].size();
            dp.assign(m, vector<int>(n, -1)); 
            return solve(0, 0, grid);
        }
    };

File: Dynamic Programming/DP on Grids/Minimum Path Sum/Memoization.java
import java.util.*;
class Solution {
    public int minPathSum(int[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        int[][] dp = new int[m][n];

        for(int[] row : dp){
            Arrays.fill(row, -1);
        }

        return recur(m, n, 0, 0, grid, dp);
    }

    public int recur(int m, int n, int i, int j, int[][] grid, int[][] dp){
        //base case 
        if(i == m-1 && j == n-1){
            return dp[i][j] = grid[i][j];
        }

        if(dp[i][j] != -1){
            return dp[i][j];
        }

        if(i == m-1){
            return dp[i][j] = recur(m,n,i,j+1,grid,dp) + grid[i][j];
        }

        if(j == n-1){
            return dp[i][j] = recur(m,n,i+1,j,grid,dp) + grid[i][j];
        }

            int right = recur(m,n,i,j+1,grid,dp) + grid[i][j];
            int down = recur(m,n,i+1,j,grid,dp) + grid[i][j];
            dp[i][j] = Math.min(right, down);

        return dp[i][j];
    }
}